snow/validators/set.go

```diff
@@ -18,92 +19,18 @@ import (
-	errDuplicateValidator = errors.New("duplicate validator")
-	errMissingValidator   = errors.New("missing validator")
+	errZeroWeight           = errors.New("weight must be non-zero")
+	errDuplicateValidator   = errors.New("duplicate validator")
+	errMissingValidator     = errors.New("missing validator")
+	errTotalWeightNotUint64 = errors.New("total weight is not a uint64")
 )
 
-// Set of validators that can be sampled
-type Set interface {
-	formatting.PrefixedStringer
-
-	// Add a new staker to the set.
-	// Returns an error if:
-	// - [weight] is 0
-	// - [nodeID] is already in the validator set
-	// - the total weight of the validator set would overflow uint64
-	// If an error is returned, the set will be unmodified.
-	Add(nodeID ids.NodeID, pk *bls.PublicKey, txID ids.ID, weight uint64) error
-
-	// AddWeight to an existing staker.
-	// Returns an error if:
-	// - [weight] is 0
-	// - [nodeID] is not already in the validator set
-	// - the total weight of the validator set would overflow uint64
-	// If an error is returned, the set will be unmodified.
-	AddWeight(nodeID ids.NodeID, weight uint64) error
-
-	// GetWeight retrieves the validator weight from the set.
-	GetWeight(ids.NodeID) uint64
-
-	// Get returns the validator tied to the specified ID.
-	Get(ids.NodeID) (*Validator, bool)
-
-	// SubsetWeight returns the sum of the weights of the validators.
-	SubsetWeight(set.Set[ids.NodeID]) uint64
-
-	// RemoveWeight from a staker. If the staker's weight becomes 0, the staker
-	// will be removed from the validator set.
-	// Returns an error if:
-	// - [weight] is 0
-	// - [nodeID] is not already in the validator set
-	// - the weight of the validator would become negative
-	// If an error is returned, the set will be unmodified.
-	RemoveWeight(nodeID ids.NodeID, weight uint64) error
-
-	// Contains returns true if there is a validator with the specified ID
-	// currently in the set.
-	Contains(ids.NodeID) bool
-
-	// Len returns the number of validators currently in the set.
-	Len() int
-
-	// Map of the validators in this set
-	Map() map[ids.NodeID]*GetValidatorOutput
-
-	// Weight returns the cumulative weight of all validators in the set.
-	Weight() uint64
-
-	// Sample returns a collection of validatorIDs, potentially with duplicates.
-	// If sampling the requested size isn't possible, an error will be returned.
-	Sample(size int) ([]ids.NodeID, error)
-
-	// When a validator's weight changes, or a validator is added/removed,
-	// this listener is called.
-	RegisterCallbackListener(SetCallbackListener)
-}
-
-type SetCallbackListener interface {
-	OnValidatorAdded(validatorID ids.NodeID, pk *bls.PublicKey, txID ids.ID, weight uint64)
-	OnValidatorRemoved(validatorID ids.NodeID, weight uint64)
-	OnValidatorWeightChanged(validatorID ids.NodeID, oldWeight, newWeight uint64)
-}
-
-// NewSet returns a new, empty set of validators.
-func NewSet() Set {
+// newSet returns a new, empty set of validators.
+func newSet() *vdrSet {
```

Comment:
Can we integrate further manager and set?
Say having a single map in Manager collecting all subnets and nodeID info?
We do something similar in the P-chain [here](https://github.com/ava-labs/avalanchego/blob/b820b46ee8d51ab1ae05da3294ff941e6f901137/vms/platformvm/state/stakers.go#L100)
This would avoid using the double locks (in manager and then in set) and it may also help us to come up with more efficient way to iterate the data
I can see the idea is of un-exporting set and keep the implementation in two levels (manager for subnets then set for a subnet nodes) is the solution with the least differences. But we may come up with more efficient solution if we integrate them further, since now set is not exported.

Reply:
Manager uses read locks currently (other than `AddStaker` + `RegisterCallback`). If we wanted to move those maps into manager, then we need to use write locks for `AddWeight, RemoveWeight`, which would lock all subnets in case another accesses the manager. 

I tried to relax current locks a bit further.